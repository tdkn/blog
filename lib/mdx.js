import React from "react";
import InlineCode from "~/components/InlineCode";
import CodeBlock from "~/components/CodeBlock";

export const components = {
  pre: CodeBlock,
  inlineCode: InlineCode
};
