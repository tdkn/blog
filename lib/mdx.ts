import path from "path";
import glob from "glob";

// POSTS_PATH is useful when you want to get the path to a specific file
export const POSTS_PATH = path.join(process.cwd(), "posts");

export const postFiles = glob
  .sync("**/*.mdx", { cwd: POSTS_PATH })
  .map((relativePath: string) => {
    const [year, slug] = relativePath.replace(/\.mdx?$/, "").split("/");

    return {
      absolutePath: path.join(POSTS_PATH, relativePath),
      relativePath,
      year,
      slug,
    };
  });
