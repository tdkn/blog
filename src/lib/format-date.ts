import { format, formatDistance, parseISO } from "date-fns";

export const formatDate = (
  dateStr: Date | string,
  formatStr = "MMMM d, yyyy",
) => {
  if (typeof dateStr === "string") {
    return format(parseISO(dateStr), formatStr);
  } else {
    return format(dateStr, formatStr);
  }
};

export const formatTimeAgo = (dateStr: Date | string) => {
  if (typeof dateStr === "string") {
    return formatDistance(parseISO(dateStr), new Date(), { addSuffix: true });
  } else {
    return formatDistance(dateStr, new Date(), { addSuffix: true });
  }
};
