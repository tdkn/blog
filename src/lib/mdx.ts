import { parseISO } from "date-fns";
import fs from "fs";
import { glob } from "glob";
import matter from "gray-matter";
import path from "path";

import type { Post } from "~/types/post";

export async function getAllPosts(): Promise<Post[]> {
  const POSTS_PATH: string = path.join(process.cwd(), "posts");
  const postFiles = await glob("**/*.mdx", { cwd: POSTS_PATH });

  const posts = postFiles.map((relativePath: string) => {
    const [year, slug] = relativePath.replace(/\.mdx?$/, "").split("/");
    const absolutePath = path.join(POSTS_PATH, relativePath);
    const source = fs.readFileSync(absolutePath);
    const { content, data: frontMatter } = matter(source);

    let date: Date = new Date();
    let deprecated: boolean = false;
    let published: boolean = true;
    let summary: string = "";
    let title: string = "";

    if (typeof frontMatter.date === "string") {
      date = parseISO(frontMatter.date);
    }

    if (typeof frontMatter.deprecated === "boolean") {
      deprecated = frontMatter.deprecated;
    }

    if (typeof frontMatter.published === "boolean") {
      published = frontMatter.published;
    }

    if (typeof frontMatter.summary === "string") {
      summary = frontMatter.summary;
    }

    if (typeof frontMatter.title === "string") {
      title = frontMatter.title;
    }

    return {
      content,
      date,
      deprecated,
      published,
      slug,
      source: "local" as const,
      summary,
      title,
      url: `/${year}/${slug}`,
      year,
    };
  });

  return posts.filter((post) => post.published);
}
