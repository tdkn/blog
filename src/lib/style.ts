import clsx, { ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const classNames = (...values: ClassValue[]) => twMerge(clsx(values));
