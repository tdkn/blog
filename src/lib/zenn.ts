import { parseISO } from "date-fns";

import type { Post, ZennArticle } from "~/types/post";

export async function getZennArticles(): Promise<{
  articlesData: ZennArticle[];
  posts: Post[];
}> {
  try {
    const response = await fetch(
      "https://zenn.dev/api/articles?username=tdkn",
      {
        next: { revalidate: 3600 },
      },
    );

    if (!response.ok) {
      console.error("Failed to fetch Zenn articles:", response.status);
      return { articlesData: [], posts: [] };
    }

    const data = (await response.json()) as { articles: ZennArticle[] };

    const posts = data.articles.map((article) => ({
      content: "",
      date: parseISO(article.published_at),
      deprecated: false,
      emoji: article.emoji,
      published: true,
      slug: article.slug,
      source: "zenn" as const,
      summary: "",
      title: article.title,
      url: `https://zenn.dev${article.path}`,
      year: parseISO(article.published_at).getFullYear().toString(),
    }));

    return { articlesData: data.articles, posts };
  } catch (error) {
    console.error("Error fetching Zenn articles:", error);
    return { articlesData: [], posts: [] };
  }
}
