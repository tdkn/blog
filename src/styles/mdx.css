:not(pre) > code {
  @apply inline-flex items-center rounded-md bg-gray-100 px-1.5 dark:bg-gray-100/10;
}

[data-rehype-pretty-code-fragment] code {
  @apply grid min-w-full;
}

[data-rehype-pretty-code-fragment] [data-line] {
  @apply px-4;
}

[data-rehype-pretty-code-fragment] [data-highlighted-chars] {
  @apply rounded-md bg-[#F1FA8C80] px-1 py-0.5;
}

[data-rehype-pretty-code-fragment] [data-highlighted-line] {
  @apply bg-[#424450];
}
