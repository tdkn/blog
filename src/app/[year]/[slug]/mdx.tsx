import { useMDXComponent } from "next-contentlayer/hooks";

import { components } from "~/components/mdx-components";
import "~/styles/mdx.css";

type MdxProps = {
  code: string;
};

export default function Mdx({ code }: MdxProps) {
  const MDXComponent = useMDXComponent(code);

  return <MDXComponent components={components} />;
}
