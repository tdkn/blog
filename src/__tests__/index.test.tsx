import { render } from "@testing-library/react";
import { posts } from "~/lib/mdx";
import Home from "~/pages/index";

describe("Home", () => {
  it("renders a Home", async () => {
    jest.spyOn(console, "error").mockImplementation();

    render(<Home posts={posts} />);

    expect(console.error).not.toHaveBeenCalled();
  });
});
