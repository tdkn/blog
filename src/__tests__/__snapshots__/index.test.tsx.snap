// Vitest Snapshot v1

exports[`Home > renders homepage unchanged 1`] = `
<div>
  <aside
    class="flex flex-row items-center text-sm"
  >
    <img
      alt="Shun Tedokon"
      class="inline-block h-16 w-16 rounded-full"
      data-nimg="1"
      decoding="async"
      height="64"
      loading="lazy"
      src="/_next/image?url=%2Favatar.jpg&w=128&q=75"
      srcset="/_next/image?url=%2Favatar.jpg&w=64&q=75 1x, /_next/image?url=%2Favatar.jpg&w=128&q=75 2x"
      style="color: transparent;"
      width="64"
    />
    <p
      class="ml-4"
    >
      Personal blog by
       
      <a
        class="dark:text-yellow-200"
        href="https://twitter.com/tdkn_"
      >
        Shun Tedokon
      </a>
      .
      <br />
      I write about design, programming, and thinking.
    </p>
  </aside>
  <main
    class="pt-8"
  >
    <article
      class="py-5"
    >
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="text-sm font-medium text-gray-500 dark:text-gray-400"
        >
          May 10, 2022
        </div>
      </div>
      <a
        class="dark:text-yellow-200"
        href="/2022/aws-cli"
      >
        <h2
          class="text-xl font-bold text-black dark:text-yellow-200"
        >
          AWS CLI を docker run 経由で使う
        </h2>
      </a>
    </article>
    <article
      class="py-5"
    >
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="text-sm font-medium text-gray-500 dark:text-gray-400"
        >
          February 24, 2021
        </div>
        <span
          class="inline-flex items-center rounded-md bg-yellow-100 px-2 py-0.5 text-xs font-medium text-yellow-800 dark:bg-yellow-100/20 dark:text-orange-500"
        >
          Deprecated
        </span>
      </div>
      <a
        class="dark:text-yellow-200"
        href="/2021/vercel-next-playwright"
      >
        <h2
          class="text-xl font-bold text-black dark:text-yellow-200"
        >
          Vercel + Next.js + PlaywrightでOGP画像をページ別に自動生成する
        </h2>
      </a>
    </article>
    <article
      class="py-5"
    >
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="text-sm font-medium text-gray-500 dark:text-gray-400"
        >
          January 21, 2021
        </div>
      </div>
      <a
        class="dark:text-yellow-200"
        href="/2021/how-to-use-yarn-patch"
      >
        <h2
          class="text-xl font-bold text-black dark:text-yellow-200"
        >
          yarn patch の使い方
        </h2>
      </a>
    </article>
    <article
      class="py-5"
    >
      <div
        class="flex items-center space-x-2"
      >
        <div
          class="text-sm font-medium text-gray-500 dark:text-gray-400"
        >
          April 9, 2020
        </div>
        <span
          class="inline-flex items-center rounded-md bg-yellow-100 px-2 py-0.5 text-xs font-medium text-yellow-800 dark:bg-yellow-100/20 dark:text-orange-500"
        >
          Deprecated
        </span>
      </div>
      <a
        class="dark:text-yellow-200"
        href="/2020/swiftui-darkmode"
      >
        <h2
          class="text-xl font-bold text-black dark:text-yellow-200"
        >
          SwiftUI のダークモードプレビュー時の不具合を回避する
        </h2>
      </a>
    </article>
  </main>
</div>
`;
